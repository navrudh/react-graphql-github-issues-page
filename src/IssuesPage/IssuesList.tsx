import * as React from 'react';

import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Issue from './Issue';

interface QueryResult {
  repository: Repository;
}

interface Repository {
  issues: Issues;
}

interface Issues {
  totalCount: number;
  edges: IssueEdge[];
}

interface IssueEdge {
  node: IssueNode;
}

export interface IssueNode {
  id: string;
  title: string;
  url: string;
  labels: IssueLabel;
}

interface IssueLabel {
  edges: LabelEdges[];
}

interface LabelEdges {
  node: LabelNode;
}

interface LabelNode {
  name: string;
  color: string;
}

const query = (repoOwner: string, repoName: string, issuesFirst: number, issueStatus: 'OPEN' | 'CLOSED') => gql`
{
  repository(owner: ${repoOwner}, name: ${repoName}) {
    issues(first: ${issuesFirst}, states: ${issueStatus}) {
      totalCount
      edges {
        node {
          id
          title
          url
          labels(first: 5) {
            edges {
              node {
                name
                color
              }
            }
          }
        }
      }
    }
  }
}

`;

interface IssuesQueryProps {
  repoOwner: string;
  repoName: string;
  issuesFirst?: number;
  issueStatus?: 'OPEN' | 'CLOSED';
}

class IssuesList extends React.Component<IssuesQueryProps> {
  private repoOwner: string;
  private repoName: string;
  private issuesFirst: number;
  private issueStatus: 'OPEN' | 'CLOSED';

  constructor(props: IssuesQueryProps) {
    super(props);
    props = props;
    this.repoOwner = props.repoOwner;
    this.repoName = props.repoName;
    this.issuesFirst = props.issuesFirst || 20;
    this.issueStatus = props.issueStatus || 'OPEN';
  }

  public render() {
    return (
      <Query query={query(this.repoOwner, this.repoName, this.issuesFirst, this.issueStatus)}>
        {({ loading, error, data }) => {
          if (loading) {
            return <p>Loading...</p>;
          }
          if (error) {
            return <p>Error :(</p>;
          }

          const queryResult: QueryResult = data;
          return queryResult.repository.issues.edges.map(({ node }) => <Issue {...node} />);
        }}
      </Query>
    );
  }
}

export default IssuesList;
