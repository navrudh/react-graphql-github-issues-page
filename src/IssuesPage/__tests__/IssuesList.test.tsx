import * as React from 'react';

import IssuesList from '../IssuesPage';
import { shallow } from 'enzyme';

describe('IssuesList', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<IssuesList />);
    expect(wrapper).toMatchSnapshot();
  });
});
