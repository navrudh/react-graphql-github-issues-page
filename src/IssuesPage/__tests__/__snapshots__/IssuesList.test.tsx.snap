// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesList renders correctly 1`] = `
<styled.div>
  <IssuesList
    cursor={null}
    firstCursor={null}
    itemsPerPage={20}
    repoName="angular"
    repoOwner="angular"
    totalCount={1}
  />
  <Query
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "owner",
                      },
                      "value": Object {
                        "block": false,
                        "kind": "StringValue",
                        "value": "angular",
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "value": Object {
                        "block": false,
                        "kind": "StringValue",
                        "value": "angular",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "repository",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [
                          Object {
                            "kind": "Argument",
                            "name": Object {
                              "kind": "Name",
                              "value": "first",
                            },
                            "value": Object {
                              "kind": "IntValue",
                              "value": "20",
                            },
                          },
                          Object {
                            "kind": "Argument",
                            "name": Object {
                              "kind": "Name",
                              "value": "after",
                            },
                            "value": Object {
                              "kind": "NullValue",
                            },
                          },
                          Object {
                            "kind": "Argument",
                            "name": Object {
                              "kind": "Name",
                              "value": "states",
                            },
                            "value": Object {
                              "kind": "EnumValue",
                              "value": "OPEN",
                            },
                          },
                        ],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "issues",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "totalCount",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "pageInfo",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "startCursor",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "endCursor",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "hasNextPage",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "hasPreviousPage",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 239,
          "start": 0,
        },
      }
    }
  />
</styled.div>
`;
